README.md
# glsl-smaa

[![npm version](https://img.shields.io/npm/v/glsl-smaa)](https://www.npmjs.com/package/glsl-smaa)
[![stability-stable](https://img.shields.io/badge/stability-stable-green.svg)](https://www.npmjs.com/package/glsl-smaa)
[![npm minzipped size](https://img.shields.io/bundlephobia/minzip/glsl-smaa)](https://www.npmjs.com/package/glsl-smaa)
[![dependencies](https://img.shields.io/david/dmnsgn/glsl-smaa)](https://github.com/dmnsgn/glsl-smaa/blob/main/package.json)
[![types](https://img.shields.io/npm/types/glsl-smaa)](https://github.com/microsoft/TypeScript)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-fa6673.svg)](https://conventionalcommits.org)
[![styled with prettier](https://img.shields.io/badge/styled_with-Prettier-f8bc45.svg?logo=prettier)](https://github.com/prettier/prettier)
[![linted with eslint](https://img.shields.io/badge/linted_with-ES_Lint-4B32C3.svg?logo=eslint)](https://github.com/eslint/eslint)
[![license](https://img.shields.io/github/license/dmnsgn/glsl-smaa)](https://github.com/dmnsgn/glsl-smaa/blob/main/LICENSE.md)

SMAA (Enhanced Subpixel Morphological Antialiasing) post-processing; WebGL (OpenGL ES 2.0) implementation with no fluff. All credit goes to the incredible work [here (iryoku)](http://www.iryoku.com/smaa/) and [there (beakbeak)](https://github.com/beakbeak/smaa-webgl).

[![paypal](https://img.shields.io/badge/donate-paypal-informational?logo=paypal)](https://paypal.me/dmnsgn)
[![coinbase](https://img.shields.io/badge/donate-coinbase-informational?logo=coinbase)](https://commerce.coinbase.com/checkout/56cbdf28-e323-48d8-9c98-7019e72c97f3)
[![twitter](https://img.shields.io/twitter/follow/dmnsgn?style=social)](https://twitter.com/dmnsgn)

## Installation

```bash
npm install glsl-smaa
```

## Usage

Three passes are required to apply the effect:

```
             |input|------------------·
                v                     |
      [ SMAA*EdgeDetection ]          |
                v                     |
            |edgesTex|                |
                v                     |
[ SMAABlendingWeightCalculation ]     |
                v                     |
            |blendTex|                |
                v                     |
  [ SMAANeighborhoodBlending ] <------·
                v
             |output|
```

You would typically set this up as part of a post-processing chain with FBOs.

See the [demo](https://dmnsgn.github.io/glsl-smaa/) and its [source](demo/index.js) for an example with [REGL](http://regl.party/).

## Attributes

Attributes required for all vertex shaders:

| Name          | Type | Description      |
| ------------- | ---- | ---------------- |
| **aPosition** | vec2 | Screen vertices. |

## Uniforms and defines

Uniforms required for all shaders:

| Name           | Type | Description           |
| -------------- | ---- | --------------------- |
| **resolution** | vec2 | The framebuffer size. |

`#define` can be preprended to the shaders to overwrite default settings. Typically:

```js
const frag = `#define SMAA_PRESET_HIGH
${PRESETS}
${SMAA_WEIGHTS_FRAG}
`;
```

### `edges.vert`

N/A

### `edges-<edge-detection-method>.frag`

You can use one of the 3 following edge detection method:

- Depth: the fastest but it may miss some edges.

- Luma: more expensive than depth edge detection, but catches visible edges that depth edge detection can miss.

- Color: the most expensive one but catches chroma-only edges.

![](https://raw.githubusercontent.com/dmnsgn/glsl-smaa/main/demo/preview-edges.gif)

Defines:

```glsl
#define SMAA_THRESHOLD 0.1
#define SMAA_LOCAL_CONTRAST_ADAPTATION_FACTOR 2.0
```

| Name                                      | Default [Range] | Description                                                                                                                                                                                                                                                                                                                           |
| ----------------------------------------- | --------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **SMAA_THRESHOLD**                        | 0.1 [0, 0.5]    | Specifies the threshold or sensitivity to edges. Lowering this value you will be able to detect more edges at the expense of performance. Range: [0, 0.5]. 0.1 is a reasonable value, and allows to catch most visible edges. 0.05 is a rather overkill value, that allows to catch 'em all.                                          |
| **SMAA_LOCAL_CONTRAST_ADAPTATION_FACTOR** | 2.0             | If there is an neighbor edge that has SMAA_LOCAL_CONTRAST_FACTOR times bigger contrast than current edge, current edge will be discarded. This allows to eliminate spurious crossing edges, and is based on the fact that, if there is too much contrast in a direction, that will hide perceptually contrast in the other neighbors. |

Uniforms:

#### `luma-edges.frag`

| Name         | Type      | Description                  |
| ------------ | --------- | ---------------------------- |
| **colorTex** | sampler2D | The input color framebuffer. |

#### `color-edges.frag`

| Name         | Type      | Description                  |
| ------------ | --------- | ---------------------------- |
| **colorTex** | sampler2D | The input color framebuffer. |

#### `depth-edges.frag`

| Name         | Type      | Description                  |
| ------------ | --------- | ---------------------------- |
| **depthTex** | sampler2D | The input depth framebuffer. |

---

### `smaa-weights.(vert|frag)`

Defines from `presets.glsl`:

```glsl
# Use presets:
#define SMAA_PRESET_LOW
#define SMAA_PRESET_MEDIUM
#define SMAA_PRESET_HIGH
#define SMAA_PRESET_ULTRA

# or individually setting:
#define SMAA_THRESHOLD 0.1
#define SMAA_MAX_SEARCH_STEPS 16
#define SMAA_MAX_SEARCH_STEPS_DIAG 8
#define SMAA_CORNER_ROUNDING 25

# and optionally disable diagonal and corner detection:
#define SMAA_DISABLE_DIAG_DETECTION
#define SMAA_DISABLE_CORNER_DETECTION
```

| Name                           | Default [Range] | Description                                                                                                                                                                                                                                                                                                              |
| ------------------------------ | --------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| **SMAA_THRESHOLD**             | 0.1 [0, 0.5]    | Should the same as for the edge pass.                                                                                                                                                                                                                                                                                    |
| **SMAA_MAX_SEARCH_STEPS**      | 16 [0, 112]     | Specifies the maximum steps performed in the horizontal/vertical pattern searches, at each side of the pixel. In number of pixels, it's actually the double. So the maximum line length perfectly handled by, for example 16, is 64 (by perfectly, we meant that longer lines won't look as good, but still antialiased. |
| **SMAA_MAX_SEARCH_STEPS_DIAG** | 8 [0, 20]       | Specifies the maximum steps performed in the diagonal pattern searches, at each side of the pixel. In this case we jump one pixel at time, instead of two.                                                                                                                                                               |
| **SMAA_CORNER_ROUNDING**       | 25 [0, 100]     | Specifies how much sharp corners will be rounded.                                                                                                                                                                                                                                                                        |

Uniforms:

| Name          | Type      | Description                                                     |
| ------------- | --------- | --------------------------------------------------------------- |
| **edgesTex**  | sampler2D | The framebuffer with edges.                                     |
| **areaTex**   | sampler2D | Precalculated textures loadable from `textures.js` base64 urls. |
| **searchTex** | sampler2D | Precalculated textures loadable from `textures.js` base64 urls. |

---

### `smaa-blend.(vert|frag)`

| Name         | Type      | Description                   |
| ------------ | --------- | ----------------------------- |
| **colorTex** | sampler2D | The input color framebuffer.  |
| **blendTex** | sampler2D | The framebuffer with weights. |

---

## License

Copyright (C) 2013 Jorge Jimenez (jorge@iryoku.com)
Copyright (C) 2013 Jose I. Echevarria (joseignacioechevarria@gmail.com)
Copyright (C) 2013 Belen Masia (bmasia@unizar.es)
Copyright (C) 2013 Fernando Navarro (fernandn@microsoft.com)
Copyright (C) 2013 Diego Gutierrez (diegog@unizar.es)

MIT. See [license file](https://github.com/dmnsgn/glsl-smaa/blob/main/LICENSE.md).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

# 2.0.0 (2021-03-31)


### Bug Fixes

* flip search and area textures ([191590a](https://github.com/dmnsgn/glsl-smaa/commit/191590ac9245d898b2a6e0ef720ce1734623fb39))
* remove round and offset defines from weights fragment shader ([650bcbf](https://github.com/dmnsgn/glsl-smaa/commit/650bcbfbccacd57afbc853f88f5db5c4d57849a6))
* update search texture ([dede139](https://github.com/dmnsgn/glsl-smaa/commit/dede1395a143c777c8c95bdf8716faeebf31fd78))


### Code Refactoring

* use ES modules ([c20d12f](https://github.com/dmnsgn/glsl-smaa/commit/c20d12fad379ac63a7f702541a7067bc2b0ec5bf))


### Features

* move presets in separate file ([028253f](https://github.com/dmnsgn/glsl-smaa/commit/028253f4f5e301d0a963ca904244d37982fdcf81))


### BREAKING CHANGES

* switch to type module


demo/commands.js
const createCube = require("primitive-cube");
const createSphere = require("primitive-sphere");
const computeNormals = require("geom-normals");
const bunnyGeometry = require("bunny");

const regl = require("./context.js");

const FBO = require("./fbos.js");

const {
  COPY_VERT,
  COPY_FRAG,
  DEPTH_FRAG,
  ENVMAP_VERT,
  ENVMAP_FRAG,
  MATERIAL_VERT,
  MATERIAL_FRAG,
  LAYER_FRAG,
  FXAA_FRAG,
  PRESETS,
  EDGES_VERT,
  LUMA_EDGES_FRAG,
  COLOR_EDGES_FRAG,
  DEPTH_EDGES_FRAG,
  SMAA_WEIGHTS_VERT,
  SMAA_WEIGHTS_FRAG,
  SMAA_BLEND_VERT,
  SMAA_BLEND_FRAG
} = require("./shaders");

// Env map
const envMap = regl({
  vert: ENVMAP_VERT,
  frag: ENVMAP_FRAG,
  uniforms: {
    uCameraView: regl.prop("uCameraView"),
    uEnvMap: regl.prop("uEnvMap")
  },
  attributes: {
    aPosition: [-4, -4, -4, 4, 8, 0]
  },
  depth: {
    enable: false,
    mask: false
  },
  count: 3
});

// Meshes
const cubeGeometry = createCube(1, 1, 1, 100, 100, 100);
const sphereGeometry = createSphere(1, { segments: 16 });

const cube = regl({
  vert: `#define UVS
  ${MATERIAL_VERT}
  `,
  frag: `#define UVS
  ${MATERIAL_FRAG}
  `,
  attributes: {
    aPosition: cubeGeometry.positions,
    aNormal: cubeGeometry.normals,
    aUv: cubeGeometry.uvs
  },
  uniforms: {
    uModelMatrix: regl.prop("uModelMatrix"),
    uModelViewMatrix: regl.prop("uModelViewMatrix"),
    uProjectionMatrix: regl.prop("uProjectionMatrix"),
    uCameraPosition: regl.prop("uCameraPosition"),
    uNormalMatrix: regl.prop("uNormalMatrix"),
    uDiffuseColor: regl.prop("uDiffuseColor"),
    uColorMap: regl.prop("uColorMap"),
    uEnvMap: regl.prop("uEnvMap")
  },
  elements: cubeGeometry.cells
});
const sphere = regl({
  vert: `#define UVS
  ${MATERIAL_VERT}
  `,
  frag: `#define UVS
  ${MATERIAL_FRAG}
  `,
  attributes: {
    aPosition: sphereGeometry.positions,
    aNormal: sphereGeometry.normals,
    aUv: sphereGeometry.uvs
  },
  uniforms: {
    uModelMatrix: regl.prop("uModelMatrix"),
    uModelViewMatrix: regl.prop("uModelViewMatrix"),
    uProjectionMatrix: regl.prop("uProjectionMatrix"),
    uCameraPosition: regl.prop("uCameraPosition"),
    uNormalMatrix: regl.prop("uNormalMatrix"),
    uDiffuseColor: regl.prop("uDiffuseColor"),
    uColorMap: regl.prop("uColorMap"),
    uEnvMap: regl.prop("uEnvMap")
  },
  elements: sphereGeometry.cells
});
const bunny = regl({
  vert: MATERIAL_VERT,
  frag: MATERIAL_FRAG,
  attributes: {
    aPosition: bunnyGeometry.positions,
    aNormal: computeNormals(bunnyGeometry.positions, bunnyGeometry.cells)
  },
  uniforms: {
    uModelMatrix: regl.prop("uModelMatrix"),
    uModelViewMatrix: regl.prop("uModelViewMatrix"),
    uProjectionMatrix: regl.prop("uProjectionMatrix"),
    uCameraPosition: regl.prop("uCameraPosition"),
    uNormalMatrix: regl.prop("uNormalMatrix"),
    uDiffuseColor: regl.prop("uDiffuseColor"),
    uColorMap: regl.prop("uColorMap"),
    uEnvMap: regl.prop("uEnvMap")
  },
  elements: bunnyGeometry.cells
});

const PRESET = `#define SMAA_PRESET_ULTRA`;

// SMAA
const SMAADepthEdges = regl({
  vert: `${PRESET}
  ${PRESETS}
  ${EDGES_VERT}`,
  frag: `${PRESET}
  ${PRESETS}
  ${DEPTH_EDGES_FRAG}`,
  attributes: {
    aPosition: [-4, -4, 4, -4, 0, 4]
  },
  uniforms: {
    resolution: [regl._gl.drawingBufferWidth, regl._gl.drawingBufferHeight],
    depthTex: FBO.depth
  },
  depth: {
    enable: false,
    mask: false
  },
  count: 3
});
const SMAALumaEdges = regl({
  vert: `${PRESET}
  ${PRESETS}
  ${EDGES_VERT}`,
  frag: `${PRESET}
  ${PRESETS}
  ${LUMA_EDGES_FRAG}`,
  attributes: {
    aPosition: [-4, -4, 4, -4, 0, 4]
  },
  uniforms: {
    resolution: [regl._gl.drawingBufferWidth, regl._gl.drawingBufferHeight],
    colorTex: FBO.SMAA
  },
  depth: {
    enable: false,
    mask: false
  },
  count: 3
});
const SMAAColorEdges = regl({
  vert: `${PRESET}
  ${PRESETS}
  ${EDGES_VERT}`,
  frag: `${PRESET}
  ${PRESETS}
  // #define SMAA_THRESHOLD 0.1
  // #define SMAA_LOCAL_CONTRAST_ADAPTATION_FACTOR 2.0
  ${COLOR_EDGES_FRAG}
  `,
  attributes: {
    aPosition: [-4, -4, 4, -4, 0, 4]
  },
  uniforms: {
    resolution: [regl._gl.drawingBufferWidth, regl._gl.drawingBufferHeight],
    colorTex: FBO.SMAA
  },
  depth: {
    enable: false,
    mask: false
  },
  count: 3
});

const SMAAWeights = regl({
  vert: `${PRESET}
  ${PRESETS}
  ${SMAA_WEIGHTS_VERT}
  `,
  frag: `${PRESET}
  ${PRESETS}
  ${SMAA_WEIGHTS_FRAG}
  `,
  attributes: {
    aPosition: [-4, -4, 4, -4, 0, 4]
  },
  uniforms: {
    resolution: [regl._gl.drawingBufferWidth, regl._gl.drawingBufferHeight],
    edgesTex: FBO.edges,
    searchTex: regl.prop("searchTex"),
    areaTex: regl.prop("areaTex")
  },
  depth: {
    enable: false,
    mask: false
  },
  count: 3
});

const SMAABlend = regl({
  vert: `${PRESET}
  ${PRESETS}
  ${SMAA_BLEND_VERT}`,
  frag: `${PRESET}
  ${PRESETS}
  ${SMAA_BLEND_FRAG}`,
  attributes: {
    aPosition: [-4, -4, 4, -4, 0, 4]
  },
  uniforms: {
    resolution: [regl._gl.drawingBufferWidth, regl._gl.drawingBufferHeight],
    colorTex: FBO.SMAA,
    blendTex: FBO.weights
  },
  depth: {
    enable: false,
    mask: false
  },
  count: 3
});

// FXAA
const FXAA = regl({
  vert: COPY_VERT,
  frag: FXAA_FRAG,
  attributes: {
    aPosition: [-4, -4, 4, -4, 0, 4]
  },
  uniforms: {
    resolution: [regl._gl.drawingBufferWidth, regl._gl.drawingBufferHeight],
    texture: FBO.color
  },
  depth: {
    enable: false,
    mask: false
  },
  count: 3
});

// SCREEN
const layers = regl({
  vert: COPY_VERT,
  frag: LAYER_FRAG,
  attributes: {
    aPosition: [-4, -4, 4, -4, 0, 4]
  },
  uniforms: {
    resolution: [regl._gl.drawingBufferWidth, regl._gl.drawingBufferHeight],
    background: regl.prop("background"),
    foreground: regl.prop("foreground")
  },
  depth: {
    enable: false,
    mask: false
  },
  count: 3
});

const toTexture = regl({
  vert: COPY_VERT,
  frag: COPY_FRAG,
  attributes: {
    aPosition: [-4, -4, 4, -4, 0, 4]
  },
  uniforms: {
    resolution: [regl._gl.drawingBufferWidth, regl._gl.drawingBufferHeight],
    texture: regl.prop("texture")
  },
  depth: {
    enable: false,
    mask: false
  },
  count: 3
});

const depth = regl({
  vert: COPY_VERT,
  frag: DEPTH_FRAG,
  attributes: {
    aPosition: [-4, -4, 4, -4, 0, 4]
  },
  uniforms: {
    near: regl.prop("near"),
    far: regl.prop("far"),
    depthTexture: regl.prop("depthTexture")
  },
  depth: {
    enable: false,
    mask: false
  },
  count: 3
});

const clear = alpha =>
  regl.clear({
    color: [0.0, 0.0, 0.0, alpha ? 0.0 : 1.0],
    depth: 1,
    stencil: 0
  });

module.exports = {
  envMap,
  cube,
  sphere,
  bunny,
  SMAADepthEdges,
  SMAALumaEdges,
  SMAAColorEdges,
  SMAAWeights,
  SMAABlend,
  FXAA,
  layers,
  toTexture,
  depth,
  clear
};
